controller/src/main/java/io/pravega/controller/server/bucket/BucketService.java

```diff
@@ -0,0 +1,275 @@
+                        }
+                        break;
+                    case ConnectivityError:
+                        log.warn("{}: StreamNotification for connectivity error", serviceType);
+                        break;
+                }
+            });
+        });
+    }
+
+    private CompletableFuture<Void> work() {
+        long time = System.currentTimeMillis();
+        QueueElement element;
+        long delayInMillis = 0L;
+        synchronized (lock) {
+            if (availableSlots > 0) {
+                element = workQueue.peek();
+                if (element != null && element.nextExecutionTimeInMillis <= time) {
+                    // Note: we can poll on queue while holding the lock because we know the element exists.
+                    element = workQueue.poll();
+                    assert element != null;
+
+                    if (!knownStreams.contains(element.getStream())) {
+                        // the stream is removed from the known set. Ignore any queue entry for this stream. 
+                        // let next cycle of process work happen immediately
+                        element = null;
+                    } else {
+                        availableSlots--;
+                    }
+                } else { 
+                    // empty priority queue or first element cannot be scheduled immediately
+                    if (element != null) {
+                        // instead of wasting compute cycles, let next run happen only when first element in the 
+                        // work queue is ready for execution. 
+                        delayInMillis = element.nextExecutionTimeInMillis - time;
+                    } else {
+                        delayInMillis = DELAY_IN_MILLIS;
+                    }
+                    element = null;
+                }
+            } else {
+                // no slots available. 
+                element = null;
+                delayInMillis = 100L;
+            }
+        }
+
+        if (element != null) {
+            Stream stream = element.getStream();
+            
+            bucketWork.doWork(stream).handle((r, e) -> {
+                long nextRun = System.currentTimeMillis() + executionPeriod.toMillis();
+                synchronized (lock) {
+                    // if known stream contains this stream, add the work back into the queue with next run time 
+                    if (knownStreams.contains(stream)) {
+                        workQueue.add(new QueueElement(stream, nextRun));
+                    }
+                    // add the slot back
+                    availableSlots++;
+                    return null;
+                }
+            });
+        } 
+        
+        // return a delayed future after which this loop is executed again. 
+        // delay is typically `0` if we have found a slot to process and a non empty queue. 
+        return Futures.delayedFuture(Duration.ofMillis(delayInMillis), executor);
```

Comment:
What happens if an item is in the queue with a long delay, and another is added with a lesser delay?